sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AsyncOffsetSeqLog.scala

```diff
@@ -0,0 +1,185 @@
+        serialize(metadata, output)
+      }.thenApply((ret: Boolean) => {
+        if (ret) {
+          batchId
+        } else {
+          throw new IllegalStateException(
+            s"Concurrent update to the log. Multiple streaming jobs detected for $batchId"
+          )
+        }
+      })
+      pendingOffsetWrites.put(batchId, future)
+      future
+    }
+
+    val lastIssuedTs = lastCommitIssuedTimestampMs.get()
+    val future: CompletableFuture[(Long, Boolean)] = {
+      if (offsetCommitIntervalMs > 0) {
+        if ((lastIssuedTs == -1) // haven't started any commits yet
+          || (lastIssuedTs + offsetCommitIntervalMs) <= clock.getTimeMillis()) {
+          issueAsyncWrite(batchId).thenApply((batchId: Long) => {
+            (batchId, true)
+          })
+        } else {
+          // just return completed future because we are not persisting this offset
+          CompletableFuture.completedFuture((batchId, false))
+        }
+      } else {
+        // offset commit interval is not enabled
+        issueAsyncWrite(batchId).thenApply((batchId: Long) => {
+          (batchId, true)
+        })
+      }
+    }
+
+    batchCache.put(batchId, metadata)
+    future
+  }
+
+  /**
+   * Adds new batch asynchronously
+   * @param batchId id of batch to write
+   * @param fn serialization function
+   * @return CompletableFuture that contains a boolean do
+   *         indicate whether the write was successfuly or not.
+   *         Future can also be completed exceptionally to indicate write errors.
+   */
+  private def addNewBatchByStreamAsync(batchId: Long)(
+      fn: OutputStream => Unit): CompletableFuture[Boolean] = {
+    val future = new CompletableFuture[Boolean]()
+    val batchMetadataFile = batchIdToPath(batchId)
+
+    if (batchCache.containsKey(batchId)) {
+      future.complete(false)
+      future
+    } else {
+      executorService.submit(new Runnable {
+        override def run(): Unit = {
+          try {
+            if (fileManager.exists(batchMetadataFile)) {
+              future.complete(false)
+            } else {
+              val start = System.currentTimeMillis()
+              write(
+                batchMetadataFile,
+                fn
+              )
+              logDebug(
+                s"Offset commit for batch ${batchId} took" +
+                s" ${System.currentTimeMillis() - start} ms to be persisted to durable storage"
+              )
+              writtenToDurableStorage.add(batchId)
+              future.complete(true)
+            }
+          } catch {
+            case e: Throwable =>
+              logError(s"Encountered error while writing batch ${batchId} to offset log", e)
+              future.completeExceptionally(e)
+          }
+        }
+      })
+      future
+    }
+  }
+
+  /**
+   * Purge entries in the offset log up to thresholdBatchId.  This method is synchronized so that
```

Comment:
Is there synchronization between the two? I don't see any synchronization here. Am I missing something?

Reply:
This log is out of date.  We don't need synchronization.  Purges always delete old entries and writes will only append new entries.  They should not conflict

Reply:
I will edit the comment